scalableminds/webknossos
# Fix volume undo for unfetched data

This PR fixes the bug that if the user draws over buckets, that aren't fetched by the backend, an undo action of that drawing will overwrite the backend data.

The problem is tackled by adding the original backend data to the volume undo once it was successfully fetched. 
Then once the actual drawing action is undone, the bucket data before the drawing action is merged together with the original backend data to not let the backend data get overwritten. Previously, the undo action would have simply taken the  bucket data before the drawing action and would not have considered the backend data that was fetched by the backend.

### URL of deployed dev instance (used for testing):
- https://fixvolumeundoforunfetcheddata.webknossos.xyz

### Steps to test:
- Open a volume / hybrid tracing without an existing segmentation layer
- Annotate a large area with one color / cell and save the tracing.
- Reload the tracing an make sure that the tracing is reloaded in the 2nd mag, not the mag with the highest resolution.
- Once the tracing is reloaded, change to another cell and draw over some of the already annotated parts in the 2nd mag.
- Then zoom in and verify that you actually overwrote the data in the lower layer.
- Zoom out again and undo your volume action. Now the whole area should only be covered with the previous cell drawing from before the reloading.
- Zoom in and verify to the highest mag and verify that the whole area is also filed there.

- Repeat the same steps on the master. In the last step, there should be holes in the annotation, as the frontend ignored the unfetched backend data of the highest mag.

### Issues:
- fixes #5256

------
